middlewares/accesslog/json_log_formatter.go

```diff
@@ -0,0 +1,235 @@
+	}
+	return mapping
+}
+
+// newJSONLogFormatter constructs a jsonLogFormatter with its four categories
+// of configuration (i.e. four sets of tuples). The core fields will be given
+// defaults if the config has not specified any.
+func newJSONLogFormatter(settings *types.AccessLog) jsonLogFormatter {
+	jlf := jsonLogFormatter{timeFormat: settings.TimeFormat}
+	ok := true
+
+	if len(settings.CoreFields) == 0 {
+		// default is to propagate all fields
+		for _, name := range defaultCoreKeys {
+			t := tuple{name, name}
+			jlf.coreMapping = append(jlf.coreMapping, t)
+		}
+	} else {
+		jlf.coreMapping = convertFieldsToMappings(settings.CoreFields)
+		ok = requireNoDuplicates(jlf.coreMapping) && ok
+		ok = validateCoreFields(jlf.coreMapping) && ok
+	}
+
+	if len(settings.RequestHeaders) > 0 {
+		jlf.requestMapping = convertFieldsToMappings(settings.RequestHeaders)
+		ok = requireNoDuplicates(jlf.requestMapping) && ok
+	}
+
+	if len(settings.OriginResponseHeaders) > 0 {
+		jlf.originMapping = convertFieldsToMappings(settings.OriginResponseHeaders)
+		ok = requireNoDuplicates(jlf.originMapping) && ok
+	}
+
+	if len(settings.DownstreamResponseHeaders) > 0 {
+		jlf.downstreamMapping = convertFieldsToMappings(settings.DownstreamResponseHeaders)
+		ok = requireNoDuplicates(jlf.downstreamMapping) && ok
+	}
+
+	if !ok {
+		exiter.Exit(1)
+	}
+
+	return jlf
+}
+
+func requireNoDuplicates(mappings []tuple) bool {
+	sourceKeys := make(map[string]struct{})
+	transKeys := make(map[string]struct{})
+	duplicates := ""
+
+	for _, m := range mappings {
+		if _, exists := sourceKeys[m.sourceKey]; exists {
+			duplicates = duplicates + " " + m.sourceKey
+		}
+		if _, exists := transKeys[m.translatedKey]; exists {
+			duplicates = duplicates + " " + m.translatedKey
+		}
+		sourceKeys[m.sourceKey] = struct{}{}
+		transKeys[m.translatedKey] = struct{}{}
+	}
+
+	if len(duplicates) > 0 {
+		log.Errorf("Duplicate access log fields:%s", duplicates)
+		return false
+	}
+	return true
+}
+
+func validateCoreFields(mappings []tuple) bool {
+	var invalidFields []string
+
+	for _, m := range mappings {
+		key := m.sourceKey
+		if _, exists := allCoreKeys[key]; !exists {
+			invalidFields = append(invalidFields, key)
+		}
+	}
+
+	if len(invalidFields) > 0 {
+		log.Errorf("Unsupported access log fields: %v", invalidFields)
+		return false
+	}
+	return true
+}
+
+//-------------------------------------------------------------------------------------------------
+
+// Exiter provides a facade for os.Exit.
+type Exiter interface {
+	Exit(code int)
+}
+
+type stdExiter struct{}
+
+func (stdExiter) Exit(code int) {
+	os.Exit(code)
+}
+
+// exiter provides a seam for testing code containing os.Exit.
+var exiter Exiter = stdExiter{}
+
+//-------------------------------------------------------------------------------------------------
+
+func (l jsonLogFormatter) Write(w io.Writer, logDataTable *LogData) error {
+
+	buf := &bytes.Buffer{}
+	buf.WriteString("{")
```

Comment:
> [...] the reason for the current implementation approach is the desire to format the JSON value (e.g. Duration) in a particular way. Because of this, simply adding everything to a map and then calling`json.Marshal` results in JSON without that formatting. I guess if we had a defined struct for the log data then we could use a custom `Marshaler`, however that would add more complexity because of the need to drop/rename certain fields based on the requested output format.

The way I see it is that we'd only need to implement custom Marshalers for those types that have custom output needs. For everything else, the default marshalling implementation already does what we need, and it's possible to mix custom and non-custom parts arbitrarily. There's also no need to construct the JSON syntax itself since Go maps and slices already provide that.

That said, talk is cheap, and I haven't implemented anything out. Though I still have my concerns, I'm okay with moving forward with the current approach if @emilevauge is too. We can always revisit our decision should the future show that this very part becomes a source of complexity.